### PR TITLE
Add Map and Mapset IPC handlers

### PR DESCRIPTION
**Pull request by IceDynamix**
Enables quaver://map/1234 and quaver://mapset/1234 links to open the map(set) ingame or download it if not present

This PR also adds the /mapset/:id API endpoint

### CODE REVIEW

**Comment by IceDynamix** [comment]
Not sure whether a dialog to download a map is necessary so I left it out for now

File: Quaver.Shared/IPC/QuaverIpcHandler.cs [hunk 1]
```diff
@@ -54,7 +62,78 @@ public static void HandleProtocolMessage(string message)
...
+                    NotificationManager.Show(NotificationLevel.Warning,
+                        "You must be in the editor to use this function!");
+            }
+            else if (message.StartsWith("map/"))
+            {
+                var game = (QuaverGame) GameBase.Game;
+                if (game.CurrentScreen.Type != QuaverScreenType.Select)
+                {
+                    NotificationManager.Show(NotificationLevel.Warning,
+                        "You must be in the song select screen to select a map!");
```
Comment (YaLTeR) [inline Quaver.Shared/IPC/QuaverIpcHandler.cs]: I guess main menu should also work for this (open song select automatically), maybe some other screens? Download?

File: Quaver.Shared/IPC/QuaverIpcHandler.cs [hunk 2]
```diff
@@ -54,7 +62,78 @@ public static void HandleProtocolMessage(string message)
...
+                    return;
+                }
+
+                var screen = game.CurrentScreen as SelectionScreen;
+
+                message = message.Replace("map/", "");
+                int mapId;
+                if (!int.TryParse(message, out mapId))
+                {
+                    NotificationManager.Show(NotificationLevel.Warning, "The provided ID was not a number!");
```
Comment (YaLTeR) [inline Quaver.Shared/IPC/QuaverIpcHandler.cs]: I'd do parsing checks before the screen check I think so people don't have to quit what they're doing just to find out the message was malformed

File: Quaver.Shared/IPC/QuaverIpcHandler.cs [hunk 3]
```diff
@@ -54,7 +62,78 @@ public static void HandleProtocolMessage(string message)
...
+                    NotificationManager.Show(NotificationLevel.Warning, "You must be logged in to download maps!");
+                    return;
+                }
+
+                // Check if the map exists online
+                var request = new APIRequestMapInformation(mapId);
+                var response = request.ExecuteRequest();
+                if (response.Status == 400 || response.Map?.MapsetId == -1)
+                {
+                    NotificationManager.Show(NotificationLevel.Warning, "The map does not exist!");
```
Comment (YaLTeR) [inline Quaver.Shared/IPC/QuaverIpcHandler.cs]: Prolly "The map does not exist on the server!" or something like that, should be clearer that it's talking about quaver online maps

File: Quaver.Shared/IPC/QuaverIpcHandler.cs [hunk 4]
```diff
@@ -54,7 +62,78 @@ public static void HandleProtocolMessage(string message)
...
+                    return;
+                } else if (response.Status != 200)
+                {
+                    NotificationManager.Show(NotificationLevel.Error, "Something happened during the request");
```
Comment (YaLTeR) [inline Quaver.Shared/IPC/QuaverIpcHandler.cs]: image

File: Quaver.Shared/IPC/QuaverIpcHandler.cs [hunk 5]
```diff
@@ -54,7 +62,78 @@ public static void HandleProtocolMessage(string message)
...
+                    var download = MapsetDownloadManager.Download(response.Map.MapsetId, response.Map.Artist, response.Map.Title);
+                    MapsetDownloadManager.OpenOnlineHub();
+
+                    // Auto import
+                    download.Completed.ValueChanged += (sender, args) =>
+                    {
+                        if (game.CurrentScreen.Type == QuaverScreenType.Select)
+                        {
+                            var selectScreen = (SelectionScreen) game.CurrentScreen;
+                            game.CurrentScreen.Exit(() => new ImportingScreen(null, true));
```
Comment (YaLTeR) [inline Quaver.Shared/IPC/QuaverIpcHandler.cs]: Is there a way to select the correct map after the set has imported? So you don't have to handle the message twice.

Comment (IceDynamix) [reply Quaver.Shared/IPC/QuaverIpcHandler.cs]: Import already automatically selects the last imported map

Comment (YaLTeR) [reply Quaver.Shared/IPC/QuaverIpcHandler.cs]: Mapset or map?

Comment (IceDynamix) [reply Quaver.Shared/IPC/QuaverIpcHandler.cs]: Right... fixed now

**Comment by Swan** [comment]
Going to open two other PRs with some rewrites to your code and also to split the map and mapset handlers since this can be done much simpler & cleaner.

**Pull request closed without merging**